### PR TITLE
Elasticsearch support

### PR DESCRIPTION
**Pull request by Tener**
This covers basic Elasticsearch support. Follow-up PR will add audit log support for queries.

Parent: https://github.com/gravitational/teleport/issues/7482

### CODE REVIEW

File: lib/srv/db/elasticsearch/test.go [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+
+type TestServer struct {
+	cfg       common.TestServerConfig
+	listener  net.Listener
+	port      string
+	tlsConfig *tls.Config
+	log       logrus.FieldLogger
+}
+
+// NewTestServer returns a new instance of a test Snowflake server.
```
Comment (AntonAM) [inline lib/srv/db/elasticsearch/test.go:48]: Not `Snowflake` ;)

File: lib/srv/db/elasticsearch/test.go [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+		return nil, trace.Wrap(err)
+	}
+
+	testServer := &TestServer{
+		cfg:       config,
+		listener:  listener,
+		port:      port,
+		tlsConfig: tlsConfig,
+		log: logrus.WithFields(logrus.Fields{
+			trace.Component: defaults.ProtocolSnowflake,
```
Comment (AntonAM) [inline lib/srv/db/elasticsearch/test.go:75]: Also shouldn't be `Snowflake`.

**Comment by AntonAM** [comment]
Nice work!

File: lib/client/db/dbcmd/dbcmd.go [hunk 3]
```diff
@@ -505,6 +508,26 @@ func (c *CLICommandBuilder) getSnowflakeCommand() *exec.Cmd {
 	return cmd
 }
 
+func (c *CLICommandBuilder) getElasticsearchCommand() *exec.Cmd {
+	args := []string{fmt.Sprintf("https://%v:%v/", c.host, c.port)}
+
+	if !c.options.noTLS {
```
[the remark this answers is not in the document]
Comment (Tener) [reply lib/client/db/dbcmd/dbcmd.go:514]: This is for consistency with the rest of codebase.

File: lib/client/db/dbcmd/dbcmd.go [hunk 4]
```diff
@@ -505,6 +508,26 @@ func (c *CLICommandBuilder) getSnowflakeCommand() *exec.Cmd {
...
+		if c.tc.InsecureSkipVerify {
+			args = append(args, "--insecure")
+		}
+
+		if c.options.caPath != "" {
+			args = append(args, []string{"--cacert", c.options.caPath}...)
+		}
+	}
+
+	return c.options.exe.Command("curl", args...)
```
[the remark this answers is not in the document]
Comment (Tener) [reply lib/client/db/dbcmd/dbcmd.go:528]: `args`: this is for consistency with the rest of codebase and I like that version better, program name is not an argument at all.

I will move the prog name into constant.

File: lib/client/db/dbcmd/dbcmd.go [hunk 5]
```diff
@@ -505,6 +508,26 @@ func (c *CLICommandBuilder) getSnowflakeCommand() *exec.Cmd {
 	return cmd
 }
 
+func (c *CLICommandBuilder) getElasticsearchCommand() *exec.Cmd {
```
[the remark this answers is not in the document]
Comment (Tener) [reply lib/client/db/dbcmd/dbcmd.go:511]: It is rather self-evident though, for internal function we have no requirement for godoc.

File: lib/srv/db/access_test.go [hunk 6]
```diff
@@ -1670,15 +1710,16 @@ func init() {
...
-		clock:       clockwork.NewFakeClockAt(time.Now()),
+		clusterName:   "root.example.com",
+		hostID:        uuid.New().String(),
+		postgres:      make(map[string]testPostgres),
+		mysql:         make(map[string]testMySQL),
+		mongo:         make(map[string]testMongoDB),
+		redis:         make(map[string]testRedis),
+		sqlServer:     make(map[string]testSQLServer),
+		snowflake:     make(map[string]testSnowflake),
+		elasticsearch: make(map[string]testElasticsearch),
```
[the remark this answers is not in the document]
Comment (Tener) [reply lib/srv/db/access_test.go:1721]: `access_test.go` arguably should be split into engine-specific test files, it is extremely long and hard to follow as is. To that end, the actual tests are in `elasticsearch_test.go`. The initialization code is common, which necessitates this change.

File: lib/srv/db/elasticsearch/engine.go [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+	}
+}
+
+func (e *Engine) HandleConnection(ctx context.Context, sessionCtx *common.Session) error {
+	if err := e.authorizeConnection(ctx); err != nil {
+		return trace.Wrap(err)
+	}
+
+	e.Audit.OnSessionStart(e.Context, sessionCtx, nil)
+	time.Sleep(time.Millisecond)
```
[the remark this answers is not in the document]
Comment (Tener) [reply lib/srv/db/elasticsearch/engine.go:111]: Hacky workaround ensuring that "session start" and "session end" have different times.

Otherwise, the backend may confusingly put "session end" before "session start" due to the clock accuracy.

File: lib/srv/db/elasticsearch/engine.go [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+
+	reqCopy.Header = req.Header.Clone()
+
+	return reqCopy, nil
+}
+
+// process reads request from connected elasticsearch client, processes the requests/responses and send data back
+// to the client.
+func (e *Engine) process(ctx context.Context, sessionCtx *common.Session, req *http.Request) error {
+	reqCopy, err := copyRequest(ctx, req, req.Body)
```
[the remark this answers is not in the document]
Comment (Tener) [reply lib/srv/db/elasticsearch/engine.go:143]: Agreed on `body` param. `copyRequest` agrees with other places in codebase.

File: lib/srv/db/elasticsearch/engine.go [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Force HTTPS usage even and update the host url.
+	reqCopy.URL.Scheme = "https"
+	reqCopy.URL.Host = sessionCtx.Database.GetURI()
```
[the remark this answers is not in the document]
Comment (Tener) [reply lib/srv/db/elasticsearch/engine.go:150]: Yeah, the method `GetURI()` returns `database connection endpoint`, which is an arbitrary string, not necessarily URI. It is pass-through from configuration and here we expect it to be db host.

File: tool/tsh/db.go [hunk 10]
```diff
@@ -563,7 +563,7 @@ func maybeStartLocalProxy(cf *CLIConf, tc *client.TeleportClient, profile *clien
 
 	// Snowflake only works in the local tunnel mode.
```
[the remark this answers is not in the document]
Comment (Tener) [reply tool/tsh/db.go:564]: True that.

**Review by AntonAM — APPROVED** [review]

File: lib/srv/db/elasticsearch/engine.go [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+	}
+
+	client := &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: tlsConfig,
+		},
+	}
+
+	// Send the request to elasticsearch API
+	resp, err := client.Do(reqCopy)
```
Comment (smallinsky) [inline lib/srv/db/elasticsearch/engine.go:162]: Does the http client needs to be created during each request handling ?

File: lib/client/db/dbcmd/dbcmd.go [hunk 12]
```diff
@@ -505,6 +510,26 @@ func (c *CLICommandBuilder) getSnowflakeCommand() *exec.Cmd {
...
```
Comment (smallinsky) [inline lib/client/db/dbcmd/dbcmd.go:513]:
```suggestion
func (c *CLICommandBuilder) getElasticSearchCommand() *exec.Cmd {
```

File: lib/client/db/dbcmd/dbcmd.go [hunk 13]
```diff
@@ -505,6 +510,26 @@ func (c *CLICommandBuilder) getSnowflakeCommand() *exec.Cmd {
...
+	args := []string{fmt.Sprintf("https://%v:%v/", c.host, c.port)}
```
Comment (smallinsky) [inline lib/client/db/dbcmd/dbcmd.go:514]: Does in case of `--tunnel`  flag the ElasticSearch client still sends the HTTPS request even if the `c.options.noTLS` option was set ?

File: lib/srv/alpnproxy/common/protocols.go [hunk 14]
```diff
@@ -91,6 +94,7 @@ var SupportedProtocols = []Protocol{
 	ProtocolRedisDB,
 	ProtocolSQLServer,
 	ProtocolSnowflake,
+	ProtocolElasticsearch,
```
Comment (smallinsky) [inline lib/srv/alpnproxy/common/protocols.go:97]:
```suggestion
	ProtocolElasticSearch,
```

File: lib/srv/db/elasticsearch_test.go [hunk 15]
```diff
@@ -0,0 +1,229 @@
...
+		require.NoError(t, err)
+		resp, err := dbConn.Ping()
+
+		require.NoError(t, err)
+		require.False(t, resp.IsError())
+		waitForEvent(t, testCtx, libevents.DatabaseSessionStartCode)
+	})
+}
+
+func withElasticsearch(name string, opts ...elasticsearch.TestServerOption) withDatabaseOption {
```
Comment (smallinsky) [inline lib/srv/db/elasticsearch_test.go:205]:
```suggestion
func withElasticSearch(name string, opts ...elasticsearch.TestServerOption) withDatabaseOption {
```

File: lib/srv/db/elasticsearch/engine.go [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+func init() {
+	common.RegisterEngine(newEngine, defaults.ProtocolElasticsearch)
+}
+
+// newEngine create new elasticsearch engine.
+func newEngine(ec common.EngineConfig) common.Engine {
+	return &Engine{EngineConfig: ec}
+}
+
+type Engine struct {
```
Comment (smallinsky) [inline lib/srv/db/elasticsearch/engine.go:49]: nit: add go doc.

File: lib/srv/db/elasticsearch/engine.go [hunk 17]
```diff
@@ -0,0 +1,207 @@
...
+	sessionCtx *common.Session
+}
+
+func (e *Engine) InitializeConnection(clientConn net.Conn, sessionCtx *common.Session) error {
+	e.clientConn = clientConn
+	e.sessionCtx = sessionCtx
+	return nil
+}
+
+func (e *Engine) SendError(err error) {
```
Comment (smallinsky) [inline lib/srv/db/elasticsearch/engine.go:64]: nit: add go doc.

File: lib/srv/db/elasticsearch/engine.go [hunk 18]
```diff
@@ -0,0 +1,207 @@
...
+		},
+	}
+
+	if err := response.Write(e.clientConn); err != nil {
+		e.Log.Errorf("elasticsearch error: %+v", trace.Unwrap(err))
+		return
+	}
+}
+
+func (e *Engine) HandleConnection(ctx context.Context, sessionCtx *common.Session) error {
```
Comment (smallinsky) [inline lib/srv/db/elasticsearch/engine.go:104]: nit: go doc

File: lib/srv/db/elasticsearch/engine.go [hunk 19]
```diff
@@ -0,0 +1,207 @@
...
+		statusCode = http.StatusUnauthorized
+	}
+
+	jsonBody, err := json.Marshal(struct {
+		Success bool   `json:"success"`
+		Message string `json:"message"`
+	}{
+		Success: false,
+		Message: err.Error(),
+	})
```
Comment (smallinsky) [inline lib/srv/db/elasticsearch/engine.go:81]: Does this is native ElasticSearch error response format ?  I in docs they are using following format
```
{
    "error" : {
        "root_cause" : [
            {
                "type" : "resource_not_found_exception",
                "reason" : "the task with id [does_not_exist] doesn't exist",
                "stack_trace": ...
            }
        ],
        "type" : "resource_not_found_exception",
        "reason" : "the task with id [does_not_exist] doesn't exist",
        "stack_trace": ...
    },
    "status": 404
}
```
Ref
https://github.com/elastic/go-elasticsearch/blob/e50e5ce23d15356e4293c95fdd3e2cf8e15f671c/typedapi/types/errorcause.go#L32

File: lib/srv/db/elasticsearch/engine.go [hunk 20]
```diff
@@ -0,0 +1,207 @@
...
+	ap, err := e.Auth.GetAuthPreference(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	mfaParams := services.AccessMFAParams{
+		Verified:       e.sessionCtx.Identity.MFAVerified != "",
+		AlwaysRequired: ap.GetRequireSessionMFA(),
+	}
+
+	dbRoleMatchers := role.DatabaseRoleMatchers(
```
Comment (smallinsky) [inline lib/srv/db/elasticsearch/engine.go:192]: DatabaseRoleMatchers will return  the default access scheme that will validate user and databaseName.

File: lib/client/db/dbcmd/dbcmd.go (hunk 12 above)
Comment (Tener) [reply lib/client/db/dbcmd/dbcmd.go:513]: Actually `Elasticsearch` seems to be the preferred capitalization: see Elastic docs, e.g. https://www.elastic.co/guide/en/elasticsearch/reference/master/index.html.

File: lib/srv/alpnproxy/common/protocols.go (hunk 14 above)
Comment (Tener) [reply lib/srv/alpnproxy/common/protocols.go:97]: `Elasticsearch` is preferred: https://github.com/gravitational/teleport/pull/15768#discussion_r958277535.

File: lib/client/db/dbcmd/dbcmd.go (hunk 12 above)
Comment (smallinsky) [reply lib/client/db/dbcmd/dbcmd.go:513]: my bad, Thought that ElasticSearch is commonly uses but after a second check the `Elasticsearch` is correct.

File: lib/srv/db/elasticsearch/engine.go (hunk 20 above)
Comment (Tener) [reply lib/srv/db/elasticsearch/engine.go:192]: True. Is this a problem?

File: lib/srv/db/elasticsearch/engine.go (hunk 16 above)
Comment (Tener) [reply lib/srv/db/elasticsearch/engine.go:52]: Done.

File: lib/srv/db/elasticsearch/engine.go (hunk 17 above)
Comment (Tener) [reply lib/srv/db/elasticsearch/engine.go:68]: Done.

File: lib/srv/db/elasticsearch/engine.go (hunk 20 above)
Comment (Tener) [reply lib/srv/db/elasticsearch/engine.go:199]: I realized we don't really support db/index restrictions for Elasticsearch, so we should probably ignore actual db name - same as for other dbs with similar limitations.

File: lib/srv/db/elasticsearch_test.go [hunk 21]
```diff
@@ -0,0 +1,207 @@
...
+
+	t.Run("session starts event", func(t *testing.T) {
+		// Connect should trigger successful session start event.
+		var err error
+
+		dbConn, proxy, err = testCtx.elasticsearchClient(ctx, "alice", "Elasticsearch", "admin")
+		require.NoError(t, err)
+		resp, err := dbConn.Ping()
+
+		require.NoError(t, err)
```
Comment (smallinsky) [inline lib/srv/db/elasticsearch_test.go:177]:
```suggestion
		resp, err := dbConn.Ping()
		require.NoError(t, err)
```

File: tool/tsh/proxy.go [hunk 22]
```diff
@@ -337,6 +337,10 @@ func onProxyCommandDB(cf *CLIConf) error {
 		return trace.BadParameter("Snowflake proxy works only in the tunnel mode. Please add --tunnel flag to enable it")
 	}
 
+	if routeToDatabase.Protocol == defaults.ProtocolElasticsearch && !cf.LocalProxyTunnel {
+		return trace.BadParameter("Elasticsearch proxy works only in the tunnel mode. Please add --tunnel flag to enable it")
+	}
```
Comment (smallinsky) [inline tool/tsh/proxy.go:342]: Nit: can we extract and combine the ` routeToDatabase.Protocol == defaults.ProtocolSnowflake && !cf.LocalProxyTunnel` and `routeToDatabase.Protocol == defaults.ProtocolElasticsearch && !cf.LocalProxyTunnel` statments into separate func.

**Review by smallinsky — APPROVED** [review]
Left 2  nit comments but otherwise LGTM.

File: tool/tsh/proxy.go (hunk 22 above)
Comment (Tener) [reply tool/tsh/proxy.go:342]: I'm struggling to find a cleaner expression without sacrificing the readability of the code or modifying the useful error message. Since this is nit I'll go ahead with the merge with the option to revisit if I can see an acceptable option.

Comment (smallinsky) [reply tool/tsh/proxy.go:342]: I was thinking to moving the `routeToDatabase.Protocol == defaults.ProtocolSnowflake && !cf.LocalProxyTunne` and `routeToDatabase.Protocol == defaults.ProtocolElasticsearch && !cf.LocalProxyTunnel` to a separate function 
Or we can just create array `requiresLocalProxy = []string{defaults.ProtocolElasticsearch, defaults.ProtocolSnowflake)` and check  
`!cf.LocalProxyTunnel && utils.Contains(routeToDatabase.Protocol,   requiresLocalProxy) `

Comment (Tener) [reply tool/tsh/proxy.go:342]: Yeah, the check is just one part though: the message is the other. We want the message to vary based on the protocol in question, so either we copy it (current approach) or construct it programmatically.

- `trace.BadParameter("%v proxy works only in the tunnel mode. Please add --tunnel flag to enable it", routeToDatabase.Protocol)` gets the case wrong. 
- `trace.BadParameter("%v proxy works only in the tunnel mode. Please add --tunnel flag to enable it", strings.Title(routeToDatabase.Protocol))` is short, but strings.Title is deprecated (even though it is fine in this context).
- The replacement of `strings.Title` takes a lot more parameters, is complex and generally hard sell for what is really capitalizing of single letter.

The shortest alternative code of semi-acceptable complexity is this:

```golang
	protocols := map[string]string{
		defaults.ProtocolSnowflake:     "Snowflake",
		defaults.ProtocolElasticsearch: "Elasticsearch",
	}

	if name, ok := protocols[routeToDatabase.Protocol]; ok && !cf.LocalProxyTunnel {
		return trace.BadParameter("%v proxy works only in the tunnel mode. Please add --tunnel flag to enable it", name)
	}
```

... but I think it is harder to read than existing code.

Thoughts?

Comment (greedy52) [reply tool/tsh/proxy.go:342]: We've added this function recently probably handy in this case
https://github.com/gravitational/teleport/blob/7744f72c6eb631791434b648ba41083b5f6d2278/lib/defaults/defaults.go#L501-L506

Comment (Tener) [reply tool/tsh/proxy.go:342]: This is neat, thanks! I'll fix the code in the next PR I have planned.
